ethcontract-generate/src/contract/types.rs

```diff
@@ -43,6 +43,9 @@ pub(crate) fn expand(kind: &ParamType) -> Result<TokenStream> {
             let size = Literal::usize_unsuffixed(*n);
             Ok(quote! { [#inner; #size] })
         }
-        ParamType::Tuple(_) => Err(anyhow!("ABIEncoderV2 is currently not supported")),
+        ParamType::Tuple(t) => {
+            let inner = t.iter().map(expand).collect::<Result<Vec<_>>>()?;
```

Comment:
As a future feature it would be super nice if this could autogenerate a named struct for the type. Do we have access to this type of info in the artefact?

Reply:
Agree it would be nice but not possible with the information we get from ethabi. According to the offical description of the json abi artifact this isn't possible because these structs-as-tuples are indistinguishable from plain tuples. In practice there is an undocumented field in the json that might hold enough information.

Reply:
I think in more recent ABI output, tuples have names. This would require `ethabi` support to parse this value.